src/main/java/de/tum/in/www1/artemis/service/ComplaintService.java

```diff
@@ -253,30 +253,54 @@ public List<Complaint> getAllComplaintsByExerciseIdAndTutorId(Long exerciseId, L
     /**
      * This function checks whether the student is allowed to submit a complaint / more feedback request or not.
      * This is allowed within the corresponding number of days set in the course after the student received the result.
-     * If the result was submitted after the assessment due date or the assessment due date is not set, the completion date of the result is checked.
-     * If the result was submitted before the assessment due date, the assessment due date is checked, as the student can only see the result after the assessment due date.
+     * This starts from the latest of these three points in time: Completion date of result, due date, assessment due date of exercise
+     *
+     * @param result the result for which a complaint should be filed
+     * @param exercise the exercise where the student wants to complain
+     * @param studentParticipation StudentParticipation the participation which might contain an individual due date
+     * @param course the course specifying the number of available days for the complaint
+     * @param type specifies if this is an actual complaint or a more feedback request
      */
-    private static void validateTimeOfComplaintOrRequestMoreFeedback(Result result, Exercise exercise, Course course, ComplaintType type) {
+    private static void validateTimeOfComplaintOrRequestMoreFeedback(Result result, Exercise exercise, StudentParticipation studentParticipation, Course course,
+            ComplaintType type) {
         int maxDays = switch (type) {
             case COMPLAINT -> course.getMaxComplaintTimeDays();
             case MORE_FEEDBACK -> course.getMaxRequestMoreFeedbackTimeDays();
         };
 
-        boolean isTimeValid;
-        if (exercise.getAssessmentDueDate() == null || result.getCompletionDate().isAfter(exercise.getAssessmentDueDate())) {
-            isTimeValid = result.getCompletionDate().isAfter(ZonedDateTime.now().minusDays(maxDays));
+        ZonedDateTime startOfComplaintTime;
+        if (exercise.getAllowComplaintsForAutomaticAssessments() || result.isRated()) {
+            ZonedDateTime relevantDueDate = studentParticipation != null && studentParticipation.getIndividualDueDate() != null ? studentParticipation.getIndividualDueDate()
+                    : exercise.getDueDate();
```

Comment:
```suggestion
            ZonedDateTime relevantDueDate = exerciseDateService.getDueDate(studentParticipation);
```
This logic already exists and should not be duplicated.

The condition of the first if-clause in `getDueDate` needs to be extended with a `if (participation != null && participation.getIndividualDueDate() != null)` in case `studentParticipation` can actually be `null` here. You can also make `getDueDate` static to avoid the dependency on the whole service.

Reply:
This is a static method, so calling the service would be a bit more tedious.
But I think I should have covered all possible null values by now

Reply:
I think it should be possible to make `ExerciseDateService::getDueDate` static as well. Then calling the service should not be a problem.